--- poldek-0.18.1/vfile/vfile.c.org	2002-11-29 16:15:35.000000000 +0100
+++ poldek-0.18.1/vfile/vfile.c	2002-11-29 16:16:24.000000000 +0100
@@ -82,18 +82,16 @@
     unsigned mod_fetch_flags;   /* passed to mod->fetch() */
 };
 
-static struct vfile_conf_s vfile_conf = { "/tmp", 0, 0 };
+static struct vfile_conf_s vfile_conf = { NULL, 0, 0 };
 
 void vfile_configure(const char *cachedir, int flags) 
 {
     int n, len = 0;
     
-    if (cachedir) {
-        vfile_conf.cachedir = n_strdup(cachedir);
-        len = strlen(vfile_conf.cachedir);
-        if (vfile_conf.cachedir[len - 1] == '/')
-            vfile_conf.cachedir[len - 1] = '\0';
-    }
+    vfile_conf.cachedir = n_strdup(cachedir);
+    len = strlen(vfile_conf.cachedir);
+    if (vfile_conf.cachedir[len - 1] == '/')
+       vfile_conf.cachedir[len - 1] = '\0';
     
     if (flags >= 0) {
         vfile_conf.flags = flags;
--- poldek-0.18.1/pkgset.c.org	2002-11-29 15:47:22.000000000 +0100
+++ poldek-0.18.1/pkgset.c	2002-11-29 15:47:52.000000000 +0100
@@ -140,7 +140,7 @@
     
     inst->rootdir = NULL;
     inst->fetchdir = NULL;
-    inst->cachedir = setup_cachedir();
+    inst->cachedir = NULL;
     inst->dumpfile = NULL;
     inst->rpmopts = NULL;
     inst->rpmacros = NULL;
--- poldek-0.18.1/main.c.org	2002-11-29 15:48:07.000000000 +0100
+++ poldek-0.18.1/main.c	2002-11-29 16:26:21.000000000 +0100
@@ -1385,6 +1385,12 @@
 
     if ((conf_get_bool(htcnf, "ftp_sysuser_as_anon_passwd", 0)))
         vfile_cnflags |= VFILE_REALUSERHOST_AS_ANONPASSWD;
+
+    // set up the cachedir
+    // dir is args.inst.cachedir if set or
+    // $TMPDIR if set or
+    // $HOME/.poldek-cache
+    args.inst.cachedir = setup_cachedir(args.inst.cachedir);
     
     vfile_configure(args.inst.cachedir, vfile_cnflags);
     
--- poldek-0.18.1/misc.c.org	2002-11-29 15:50:00.000000000 +0100
+++ poldek-0.18.1/misc.c	2002-11-29 16:08:47.000000000 +0100
@@ -138,11 +138,17 @@
     
 
 
-const char *setup_cachedir(void) 
+const char *setup_cachedir(const char *confvalue) 
 {
     struct passwd *pw;
     char *dir, *default_dn = ".poldek-cache";
 
+    if ((confvalue) && vf_valid_path(confvalue) && is_rwxdir(confvalue)) {
+        return n_strdup(confvalue);
+    }else{
+        logn(LOGWARN, _("configured cachedir is invalid, fallback to default"));
+    }
+
     if ((dir = getenv("TMPDIR")) && vf_valid_path(dir))
         return n_strdup(dir);
     
--- poldek-0.18.1/misc.h.org	2002-11-29 16:07:45.000000000 +0100
+++ poldek-0.18.1/misc.h	2002-11-29 16:08:20.000000000 +0100
@@ -30,7 +30,7 @@
   Returns $TMPDIR or "/tmp" if $TMPDIR isn't set.
   Returned dir always begin with '/'
 */
-const char *setup_cachedir(void);
+const char *setup_cachedir(const char *confvalue);
 const char *tmpdir(void);
 
 const char *expand_env_vars(char *dest, int size, const char *str);
